Syncing of Pubs in Tv3 Legacy Mode

spficklin opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- See our Contribution Guidelines here:
          https://github.com/tripal/tripal/blob/7.x-3.x/CONTRIBUTING.md -->
          
<!--- If it fixes an open issue, please add the issue link below. -->
Issue #577 

## Type(s) of Change(s)
<!--- What types of changes does your code introduce? 
         Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API-specific change (fix or addition to an API function)
- [ ] Updates documentation (inline or markdown files)

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
The original intent for this was to fix the problem described in issue #577.  The problem occurs when Tv3 is run in legacy mode, all nodes are converted to entities but the site is still using the Tv2 templates and therefore the legacy modules are still enabled.  When importing new publications Tripal detects that the tripal_pub legacy module is enabled and  automatically tries to sync new publications as nodes.   It should not do that.    

This PR fixes the problem in the following way:
1.  When importing publications via the drush command, they are no longer automatically "syncing". Now there are new arguments that allow the user to sync (nodes) or publish (entities) . 
2.  If the user clicks the "import" link on the Chado Pub Bulk Importer page then it will automatically publish as entities (never as nodes). 

Also, this issue was a bit of a rabbit hole for the following reasons:
1.  There was duplicated code for importing pubs in multiple functions. I consolidated and cleaned up this code.
2.  The job logging wasn't working for importing pubs.  So, I fixed that, but  if importing pubs is run via a Drush command I had to duplicate log messages for the terminal.  To cut down on the amount of code I added a new argument to the tripal_report_error() function that let me pass in a Tripaljob object and that function would log to the terminal and to the job if one was available.
3.  For some reason the tripal_pub and tripal_contact modules still had some function calls in the install files that were moved over to the tripal_chado module and when I enabled them I had problems.  When testing my fixes enabling these modules failed.   I am not certain why non one who is running in legacy mode didn't encounter these problems... Maybe they did and resolved them... don't know. But I fixed it.

## Testing?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
<!--- Reviewers will use this section to test the submission! -->

Perform steps 1-3 prior to pulling the branch for this PR.

1. To test this you will need to enable the tripal_pub legacy module (will enable all dependencies as well).
2.  Next go to Tripal > Data Loaders > Chado Publication Bulk Importer and create an importer if you don't already have one.  I found that an easy one to test with few publications is to set the number of days to 30, and the search term to 'Citrus'.  It returns just a few publications.  If you have some importers already you might want to deactivate them temporarily to keep this test from taking too long.
3.  Now try to run the importer using drush.  Notice it will try to sync the publications.
```
  drush trp-import-pubs --username=[username] --update=Y
``` 
4.  Pull the branch for this PR.
5.  Repeat the drush command from step 3.   It should now try to publish instead of syncing.
6.  Now run this drush command. It should try to sync but not publish
```
  drush trp-import-pubs --username=[username] --update=Y --sync=Y --publish=N
```
7.  Now run this drush command. It should try to both sync and publish.
```
  drush trp-import-pubs --username=[username] --update=Y --sync=Y --publish=Y
```
8.  Now run the following drush command.  It will neither try to sync or publish.
```
  drush trp-import-pubs --username=[username] --update=Y --sync=N --publish=N
```
9.  Now run this drush command and play with the various combinations of --sync and --publish to check all combinations like steps 6-8 above.
```
  drush trp-import-pubs --username=[username] --dbxref=PMID:24163125  --update=Y --sync=N --publish=Y
```
10.  Now go to the Tripal > Data Loaders > Chado Publication Bulk Importer and click the link to  "Import Pubs" next to the importer (choose one that is not disabled and that runs quickly).  It will submit a job that you can run with the provided drush command.  It should try to publish.   
11.  Go to the Tripal > Jobs page and click on the "Import Publications" link for the job you just ran.  You should see a page with the job details and the job log.   Open the Job logs and make sure you see the logs.

## Screenshots (if appropriate):

## Additional Notes (if any):
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

spficklin committed:
Working on issue #577

spficklin committed:
Fixed other drush commands

spficklin commented:
If you review this PR, please consider also reviewing the other one related to pub importing:  PR #585.  It should be really easy to do once you've gotten through this one.

bradfordcondon commented on legacy/tripal_contact/tripal_contact.install:
```diff
@@ -52,12 +52,12 @@ function tripal_contact_requirements($phase) {
-  $obo_id = tripal_insert_obo('Tripal Contacts', $obo_path);
-  tripal_submit_obo_job(array('obo_id' => $obo_id));
+  //$obo_path = '{tripal_contact}/files/tcontact.obo';
+  //$obo_id = tripal_insert_obo('Tripal Contacts', $obo_path);
```
is there a good reason to leave this in commented out?  I'd prefer to remove if possible.

(here and all commented out code below)

spficklin replied:
Yes, I can remove that.  I'll do that and commit a fix.

spficklin replied:
Okay, I removed it as well as some other commented out sections.

spficklin committed:
Cleaned up commented out code

almasaeed2010 commented:
Unfortunately, I am unable to test this branch because tripal_pub fails to install on my site:

```shell
drush en -y tripal_pub
The following extensions will be enabled: tripal_pub, tripal_contact, tripal_db, tripal_cv, tripal_core
Do you really want to continue? (y/n): y
WD php: Error: Call to undefined function tripal_contact_add_custom_tables() in tripal_contact_install() (line 55 of                          [error]
/Users/Almsaeed/Work/DevSites/Tripal/sites/all/modules/tripal/legacy/tripal_contact/tripal_contact.install).
Error: Call to undefined function tripal_contact_add_custom_tables() in tripal_contact_install() (line 55 of /Users/Almsaeed/Work/DevSites/Tripal/sites/all/modules/tripal/legacy/tripal_contact/tripal_contact.install).
Drush command terminated abnormally due to an unrecoverable error. 
```

almasaeed2010 commented:
Now that I have tripal_pub installed, I go to test this but I am getting the following error:

```
[site http://default] [TRIPAL ERROR] [TRIPAL_CONTACT] Could not add the contact property 'Surname'
[site http://default] [TRIPAL ERROR] [PUB_IMPORT] Cannot add property 'Publication Dbxref' to publication. Skipping.
```

Is there a step missing? Do I need to add a field `Surname` as a chado property field to the Contact bundle?

Thanks!

almasaeed2010 commented:
The funny thing is that I do have `Publication Dbxref` chado prop field in the bundle:

<img width="1362" alt="screen shot 2018-08-29 at 9 07 49 am" src="https://user-images.githubusercontent.com/1512664/44789693-41e24b80-ab6b-11e8-8604-9d2b8cea80c9.png">

And it's using the correct vocab as far as I can tell.

almasaeed2010 commented:
So, although it reported at the end that 0 records were published, I still got 68 publication entities published.

Examples:
<img width="1363" alt="screen shot 2018-08-29 at 9 23 49 am" src="https://user-images.githubusercontent.com/1512664/44790699-f1b8b880-ab6d-11e8-96f8-158b78a1eedf.png">


The other thing is when running  `drush trp-import-pubs --username=admin --update=Y --sync=N --publish=N` It shouldn’t attempt to publish but it does according to the messages:

```
INFO (PUB_IMPORT): Publishing publications with Drupal...
array_flip(): Can only flip STRING and INTEGER values! entity.controller.inc:219                                                                                       [warning]
INFO (PUBLISH_RECORDS): There are 0 records to publish.
INFO (PUBLISH_RECORDS): Succesfully published 0 Publication record(s).
INFO (PUB_IMPORT): Done.
```

Thanks!

bradfordcondon commented:
I also get ```[site http://default] [TRIPAL ERROR] [PUB_IMPORT] Cannot add property 'Publication Dbxref' to publication. Skipping.``` when running `  drush trp-import-pubs --username=[username] --update=Y`

bradfordcondon commented:
when i run the import-pubs job... whats the deal with this?

```
drush trp-import-pubs --username=admin --dbxref=PMID:24163125  -update=Y --sync=N --publish=Y

Cache HIT cid: 8.1.8-commandfiles-0-44eb5603e5a71e770c6cc3a6849fef85 [0.03 sec, 3.37 MB]                                            [debug]
Bootstrap to phase 0. [0.11 sec, 7.48 MB]                                                                                       [bootstrap]
Bootstrap to phase 1. [0.11 sec, 7.48 MB]                                                                                       [bootstrap]
Drush bootstrap phase : bootstrap_drupal_root() [0.11 sec, 7.83 MB]                                                             [bootstrap]
Initialized Drupal 7.59 root directory at /Users/chet/UTK/tripal [0.12 sec, 7.83 MB]                                            [bootstrap]
Find command files for phase 1 (max=1) [0.12 sec, 6.65 MB]                                                                          [debug]
Cache HIT cid: 8.1.8-commandfiles-1-dde52f179f1719d2e7bcd1ca94a63522 [0.12 sec, 6.65 MB]                                            [debug]
Cache HIT cid: 8.1.8-annotationfiles-1-5a8cb4bc9094ac5f68d9b1fa5b1ece36 [0.12 sec, 6.65 MB]                                         [debug]
Bootstrap to phase 5. [0.13 sec, 7.84 MB]                                                                                       [bootstrap]
Drush bootstrap phase : bootstrap_drupal_site() [0.16 sec, 7.86 MB]                                                             [bootstrap]
Initialized Drupal site default at sites/default [0.16 sec, 7.86 MB]                                                            [bootstrap]
Find command files for phase 2 (max=5) [0.17 sec, 6.67 MB]                                                                          [debug]
Drush bootstrap phase : bootstrap_drupal_configuration() [0.17 sec, 6.67 MB]                                                    [bootstrap]
Find command files for phase 3 (max=5) [0.17 sec, 6.69 MB]                                                                          [debug]
sql-query: SELECT 1; [0.17 sec, 6.74 MB]                                                                                           [notice]
Executing: PGPASSFILE=/private/tmp/drush_8dxbGW psql -q --dbname=drupal --host=localhost --port=5432 --username=drupal  --no-align --field-separator="	" --pset tuples_only=on --file /private/tmp/drush_yAaFJc
sql-query: SELECT tablename FROM pg_tables WHERE schemaname='public'; [0.2 sec, 6.75 MB]                                           [notice]
Executing: PGPASSFILE=/private/tmp/drush_imlNWE psql -q --dbname=drupal --host=localhost --port=5432 --username=drupal  --no-align --field-separator="	" --pset tuples_only=on --file /private/tmp/drush_ffVOzI
Drush bootstrap phase : bootstrap_drupal_database() [0.23 sec, 6.76 MB]                                                         [bootstrap]
Successfully connected to the Drupal database. [0.24 sec, 7.12 MB]                                                              [bootstrap]
Find command files for phase 4 (max=5) [0.24 sec, 7.12 MB]                                                                          [debug]
Drush bootstrap phase : bootstrap_drupal_full() [0.24 sec, 7.12 MB]                                                             [bootstrap]
Cache HIT cid: 8.1.8-install_profile-66ecfeb9791a023150773849f1550c5d [0.47 sec, 30.67 MB]                                          [debug]
Find command files for phase 5 (max=5) [0.47 sec, 30.67 MB]                                                                         [debug]
Cache HIT cid: 8.1.8-install_profile-66ecfeb9791a023150773849f1550c5d [0.47 sec, 30.67 MB]                                          [debug]
Cache HIT cid: 8.1.8-commandfiles-5-4829dc6049a04d0467d9e39ec3a2a121 [0.47 sec, 30.68 MB]                                           [debug]
Cache HIT cid: 8.1.8-annotationfiles-5-01d9724a39e1b765d50256a29364258f [0.47 sec, 31.12 MB]                                        [debug]
Bootstrap to phase 6. [0.49 sec, 32.58 MB]                                                                                      [bootstrap]
Drush bootstrap phase : bootstrap_drupal_login() [0.49 sec, 32.58 MB]                                                           [bootstrap]
array_flip(): Can only flip STRING and INTEGER values! entity.inc:175 [0.49 sec, 32.91 MB]                                      [warning]
Could not login with user account `1'. [0.49 sec, 32.91 MB]                                                                     [error]
Find command files for phase 6 (max=6) [0.5 sec, 31.45 MB]                                                                          [debug]
The drush command 'trp-import-pubs' could not be executed. [0.5 sec, 31.45 MB]
```


Anyway, hte meat of your PR looks good to me, although as abdullah points out it tries to publish even if publish = N

spficklin committed:
Merge branch '7.x-3.x' into 577-tv3-pub_syncing

spficklin committed:
Fixed bugs from PR review

spficklin commented:
Okay.  thank you both @almasaeed2010 and @bradfordcondon for reviewing this.  I apologize for the issues.  I've just committed a few fixes.  Can you try again?  

Here's the low-down on what went wrong:

1) @almasaeed2010 there was problems installing the legacy tripal_pub module.   Those have been addressed in another PR so I think we're good there.
2)  @almasaeed2010 and @bradfordcondon The error about "Cannot add property 'Publication Dbxref'" was due to the fact that we don't clean out that one property from the pubprop table on an update. So when it tried to insert it couldn't because it was already there. All other properties get cleared out on an update.   I added a statement to ignore that property on an update.
3)  @almasaeed2010 The "Cannot add property 'Publication Dbxref'" warning was a bit misleading it wasn't talking about TripalField properties but property records in the pubprop table. I fixed that error message to be more clear if it every comes up again.
4) @bradfordcondon that nonsense with those bootstrap messages was due to the fact that I accidentally wrote the testing command incorrectly.  If you supply the "update" argument with one dash you get that gobbly goop.  If you provide it with two dashes it works fine. I fixed the test code above to have two dashes.
5)  @almasaeed2010 if it says you have 0 records published it's probably that you already had those publications published so even though it updated the publications in Chado it didn't need to republish them.  Any new publications that you didn't have in Chado would be published.
6)  @almasaeed2010 there was in fact a mistake in the code that allowed for publishing even if you set both --publish and --sync to 'N'.  I fixed that.
7) @almasaeed2010 and @bradfordcondon. The error message 'Can only flip STRING and INTEGER' was caused by a typo in the loading of bundle (forget exactly where). I fixed that.

almasaeed2010 commented:
Hello @spficklin 

I tested all the commands. Everything worked great and data is published correctly when the parameters are set! There is a little HTML error though in the job view. I am not sure if it was introduced in this PR but I see it in this branch. 

This line: https://github.com/tripal/tripal/blob/4229d8b98f4e7aa2a3882a453a7574a618e93845/tripal/includes/tripal.jobs.inc#L323

seems to remove the last 2 characters in the links string. This is causing the entire page to render as a link. It removes the last 2 characters because it needs to get rid of the last ` | `, which doesn't exist for the execute link:
https://github.com/tripal/tripal/blob/4229d8b98f4e7aa2a3882a453a7574a618e93845/tripal/includes/tripal.jobs.inc#L286

The fix then is to change the above line to:
```
$links .= l('Execute this job', "admin/tripal/tripal_jobs/execute/".$job->job_id) . ' | ';
```

Screenshot of the problem:
<img width="861" alt="screen shot 2018-08-30 at 5 17 32 pm" src="https://user-images.githubusercontent.com/1512664/44879795-ab9c4b80-ac78-11e8-9c79-5cb95d1bc279.png">

**Note** that this issue only happens when the job has not been started yet.

Thanks

spficklin committed:
Fixed job display issue

spficklin commented:
Hi @almasaeed2010 that's very odd what's going on with the job view. I wasn't able to duplicate that behavior.  But I didn't like the way those links were being made. So I rewrote that. They now show up as "action links".  That should fix the problem for you.

![image](https://user-images.githubusercontent.com/1719352/44884762-2877f700-ac72-11e8-86bf-97907e051a0a.png)

bradfordcondon commented:
all importer commands now running as expected and error free, thanks @spficklin 

my job view looks fine. im approving.

almasaeed2010 commented:
Great! Thanks @spficklin

spficklin commented:
Thanks @bradfordcondon and @almasaeed2010 !
